### PR TITLE
Add circle-crop variation to Image block.

### PR DESCRIPTION
**Pull request by jasmussen**
Props @phpbits for the technical implementation.

This PR adds a variation to the Image block, "Circle Crop". What that means is you can choose a style variation for the image, which rounds all 4 corners. When this is appliedto a square image, that means a perfect circle. When it is applied to a rectangle, it means a pill-shape.

In past attempts to fix this, it was either an oval crop, or it required a large refactor to be able to make a rectangular image into a square, in CSS only. But with @phpbits clever trick — using a pixel value for the border radius — the we can actually implement this in a simple way that is still visually appealing even for rectangles.

GIF:

circle crop

---

On a separate note, I did look into sort of a "hack" to create perfect circles from rectangular images. Since you can manually type in the dimensions of an image in the block inspector, you can _force_ a square image even from a rectangle. If you combine that with `object-fit: cover;`, it means that the image won't be stretched. 

However, every modern WordPress theme has the following CSS bundled along, to make images responsive:

```
img {
	max-width: 100%;
	height: auto;
}
```

This reminded me that unless the width of the image is fluid, it won't scale to mobile. So the hack unfortunately won't work.

### CODE REVIEW

**Comment by jasmussen** [comment]
Fixes #16130.

File: packages/block-library/src/image/style.scss
Comment (kjellr) [inline packages/block-library/src/image/style.scss]:
```suggestion
	border-radius: 9999px;
	
	
	// If a browser supports it, we will switch to using a circular SVG mask here.
	@supports (mask-image: none) or (-webkit-mask-image: none) {
		border-radius: none;
		mask-image: url('data:image/svg+xml;utf8,<svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="50"/></svg>');
		-webkit-mask-image: url('data:image/svg+xml;utf8,<svg viewBox="0 0 100 100" xmlns="http://www.w3.org/2000/svg"><circle cx="50" cy="50" r="50"/></svg>');
		mask-mode: alpha;
		-webkit-mask-mode: alpha;
		mask-repeat: no-repeat;
		-webkit-mask-repeat: no-repeat;
		mask-size: contain;
		-webkit-mask-size: contain;
		mask-position: center;
		-webkit-mask-position: center;
	}
```

^ So this is essentially all we'd need to add to implement the idea proposed over in https://github.com/WordPress/gutenberg/issues/16130#issuecomment-509729977. The only issue is that this fails our linter, because I think we're not supposed to use both single and double quotes inside of a `url()` value. 😩 

We could convert this SVG to `base64` instead, but I know it's inadvisable to have long obscure strings in the code. OR, we could host the SVG somewhere else in Gutenberg, and call it here, but that seems weird too. Any ideas?

Comment (jasmussen) [reply packages/block-library/src/image/style.scss]: > The only issue is that this fails our linter, because I think we're not supposed to use both single and double quotes inside of a url() value. 😩

There's a CSS comment you can write to disable the linter for the next line. It exists for this purpose exactly: to disable the linter for completely valid edgecases. I can't recall the exact comment to write, but if you combine it with a comment explaining why it's being disabled, I think that's totally fine. 

Btw, I don't know if we need the -webkit- properties — postCSS is supposed to add them automatically. It might not, if the property is exotic enough, but in general it should work.

Comment (kjellr) [reply packages/block-library/src/image/style.scss]: > There's a CSS comment you can write to disable the linter for the next line. It exists for this purpose exactly: to disable the linter for completely valid edgecases.

Btw, I don't know if we need the -webkit- properties — postCSS is supposed to add them automatically. It might not, if the property is exotic enough, but in general it should work.

Good call on both counts! Just pushed these changes in c6e56ca. 👍

**Comment by kjellr** [comment]
Here's an updated GIF, now that c6e56ca is included. 

circle

**Comment by jasmussen** [comment]
I dig that. The floating resize handle actually sort of addresses the margin thing. 

I wonder if we should include the `border-radius: 9999px` version as a "Rounded" variation also? Or does that get messy?

File: packages/block-library/src/style.scss
Comment (kjellr) [inline packages/block-library/src/style.scss]: @jasmussen Is this commenting out intentional?

**Comment by kjellr** [comment]
> I wonder if we should include the border-radius: 9999px version as a "Rounded" variation also? Or does that get messy?

I think we should keep it simple and just introduce the circle for now. 

> And yes, "Circle Mask" is better!

Updated in 7418ec2

File: packages/block-library/src/style.scss
Comment (jasmussen) [reply packages/block-library/src/style.scss]: Not intentional. height: auto; should remain, good catch.

**Comment by jasmussen** [comment]
I'm digging it! Thanks for taking this over Kjell! I would suspect this is ready for review?

**Comment by kjellr** [comment]
> I'm digging it! Thanks for taking this over Kjell! I would suspect this is ready for review?

Yep! This should be ready for review.

**Review by senadir — APPROVED** [review]
LGTM, I don't see a reason to block this.

**Comment by jasmussen** [comment]
Nice, @youknowriad and @mapk can you take a final look?

**Comment by youknowriad** [comment]
There's a very small issue when resizing as the resizing handles can appear a bit far from the circle.

The other probably more concerning issue is the fact that the size of the image container doesn't adapt properly which means there's a lot of extra whitespace. It feels bad especially for vertical images or when you left/right align the image.

**Comment by jasmussen** [comment]
That was the reason for the pill shape initially in the pr, that way the handles stay attached. But then it's not a circle. I have a slight preference for that, for this reason, but not a strong one. The circle crop is clever though.

**Comment by youknowriad** [comment]
isn't there a way to "crop" the figure maybe (container of the image) to be a square once the circle is applied?

**Comment by jasmussen** [comment]
I tried very hard to do that back in the day in an old branch, try/circle-crop. I just couldn't get it to work reliably.

However as mentioned in this pr, you can choose a square crop from the sidebar.

**Comment by youknowriad** [comment]
Anyway, I don't feel like these concerns are strong enough to block this PR. Let's merge and see the feedback we'll have.

**Comment by mapk** [comment]
Thanks for the ping, @jasmussen. I just tested this and ran into a little issue. It might be beyond this PR though and related to this one: https://github.com/WordPress/gutenberg/issues/17174

If I want to change the styling after resizing the image, I have to click the Image block again to return focus, and then also open the "Styles" accordion panel in the sidebar again. Everytime the Image block loses focus that panel closes. Can we keep it open? Should we?

**Comment by senadir** [comment]
> If I want to change the styling after resizing the image, I have to click the Image block again to return focus, and then also open the "Styles" accordion panel in the sidebar again. Everytime the Image block loses focus that panel closes. Can we keep it open? Should we?

This is now fixed #17175, since everyone is all set with this PR, I will give it a last check before finally merging it.

**Pull request merged**
